### ISSUE TITLE
A `ReadOnlyTransaction` error within an inner transaction doesn't rollback the savepoint

### ISSUE THREAD

**Issue by jtgeibel** [diesel-rs/diesel#3470]
### Versions

- **Rust:** 1.66
- **Diesel:** 2.0.2
- **Database:** PostgreSQL 14.6
- **Operating System:** Ubuntu 22.10

### Feature Flags

- **diesel:** ["postgres"]

## Problem Description

Starting with diesel `2.0.0-rc.1` (commit 53bad857f30f2f32c14ae295487a88da278fb2fc of #3211) when within a nested transaction, if the inner transaction aborts because it attempts to mutate within a read-only transaction, then the connection is left in a bad state even outside of the inner transaction. Any further commands result in `current transaction is aborted, commands ignored until end of transaction block` until the outermost transaction is also rolled back. It appears that the SAVEPOINT for the inner transaction is not actually rolled back, as the equivalent series of SQL commands (including a `ROLLBACK TO SAVEPOINT ...`) does not result in such an error in `psql`.

### What are you trying to accomplish?

I am trying to port crates.io to diesel 2.0. Our tests are run within an outer transaction that rolls back when the test completes. In production the code will continue to work as expected as this code path has just the one transaction, but the additional transaction wrapping the entire test causes this test to fail.

### What is the expected output?

It is expected that if a nested transaction encounters a `ReadOnlyTransaction` error then it should still rollback to the appropriate SAVEPOINT so that future read-only commands (such as SELECT) can be executed.

### What is the actual output?

The observed output is an error stating `current transaction is aborted, commands ignored until end of transaction block`.

### Are you seeing any additional errors?

No additional errors are seen.

### Steps to reproduce

Create a simple test table with `cargo run` and then run `cargo test`. In diesel releases starting with `2.0.0-rc.1` the 2nd test fails with `Error: DatabaseError(Unknown, "current transaction is aborted, commands ignored until end of transaction block")`.

```rust
use diesel::{prelude::*, sql_query};

const DATABASE_URL: &str = "postgres:///";

fn connect() -> PgConnection {
    PgConnection::establish(DATABASE_URL).unwrap()
}

fn main() {
    sql_query("CREATE TABLE rollback_test (id INT PRIMARY KEY, value INT NOT NULL)")
        .execute(&mut connect())
        .unwrap();
}

#[cfg(test)]
mod tests {
    use super::*;
    use diesel::update;

    diesel::table! {
        rollback_test (id) {
            id -> Int4,
            value -> Int4,
        }
    }

    fn update_with_fallback_query(conn: &mut PgConnection) -> QueryResult<()> {
        conn.transaction(|conn| {
            sql_query("SET TRANSACTION READ ONLY").execute(conn)?;
            dbg!(update(rollback_test::table)
                .set(rollback_test::value.eq(0))
                .execute(conn))
        })
        .or_else(|_| sql_query("SELECT 1").execute(conn))
        .map(|_| ())
    }

    #[test]
    fn test_rollback_of_readonly() -> QueryResult<()> {
        update_with_fallback_query(&mut connect())
    }

    /// Test that start failing at rev="53bad857f30f2f32c14ae295487a88da278fb2fc"
    #[test]
    fn test_rollback_of_readonly_within_outer_transaction() -> QueryResult<()> {
        connect().transaction(|conn| update_with_fallback_query(conn))
    }
}
```

For comparison, here are similar SQL commands run via `psql`:

```sql
BEGIN;
SAVEPOINT rollback_test;
SET TRANSACTION READ ONLY;
UPDATE rollback_test SET value=1;
-- ERROR:  cannot execute UPDATE in a read-only transaction
SELECT 1;
-- ERROR:  current transaction is aborted, commands ignored until end of transaction block
ROLLBACK TO SAVEPOINT rollback_test;
SELECT 1;
-- No longer an error, as we've left the inner transaction.
-- So diesel must not be issuing the expected savepoint rollback.
COMMIT;
```

## Checklist

- [x] I have already looked over the [issue tracker](https://github.com/diesel-rs/diesel/issues)  and the [disussion forum](https://github.com/diesel-rs/diesel/discussions)  for similar possible closed issues.
- [x] This issue can be reproduced on Rust's stable channel. (Your issue will be
  closed if this is not the case)
- [x] This issue can be reproduced without requiring a third party crate

**Comment by weiznich** [comment]
Thanks for opening this issue and providing a reproduction. I will try to find a solution as soon as possible.

**Comment by weiznich** [comment]
The following post is just a dump of things that I've tried:

* Diesel executes the following commands for your testcase :
```sql
BEGIN;
SAVEPOINT diesel_savepoint_1;
SET TRANSACTION READ ONLY;
update rollback_test SET value=1;
-- ERROR: readonly transaction
SELECT 1; 
-- ERROR: current transaction is aborted, commands ignored until end of transaction block
```
Executing these commands in psql reproduces the same error. The issue is caused by the missing `ROLLBACK TO SAVEPOINT` command. That changed in the referenced commit.

* For the postgres connection we use `libpq` to get the current transaction state. `libpq` reports that the transaction state is in an error state [here](https://github.com/diesel-rs/diesel/blob/master/diesel/src/pg/connection/mod.rs#L243). This causes us to mark the connection as requiring a top level rollback and skipping all `ROLLBACK TO SAVEPOINT …` commands. The main motivation for this is to correctly handle transactions that fail due to serialization errors. 

That all written: I'm not 100% sure yet how to fix that. It might be worth to just try to perform the `ROLLBACK TO SAVEPOINT` nevertheless and ignore potential errors. 

cc @Ten0 as author of the relevant commit

**Comment by Ten0** [comment]
Pretty raw thoughts - couldn't investigate in detail on some points below:

To give more context about this particular change, Diesel 1 had poor tracking of transactions failing, which sometimes led to Diesel thinking the depth was a certain level when in fact it was at a different level (#2123). This initially was fixed by #2935, of which however the interface was not ideal (made matching on errors difficult, backend-specific code was polluting the general code, there were a few bugs left)
#3211 improves all of these points before the 2.0 release, and notably simplifies it and makes it more reliable by using the information sent back by the postgres server about the transaction state instead of trying to guess it from what the last query was and its result. 

So overall the current implementation is probably the safest way to fix an improper behavior we used to have where we would consider as not broken stuff that was broken, but that comes at the cost of apparently over-considering-broken stuff that could in theory be recovered in a more fine-grained manner. 

I can't imagine that serialization failures would be always fixed by savepoint rollback, so it looks like we shouldn't take this libpq status as "you may rollback just one level and you should be good": typically it instead means "you should rollback top-level" (which is the case for serialization failures - or can savepoints recover some of them?), or it's a programmatic misuse (your test case IIUC). 

So it looks like we have essentially two solutions:
- Some specific (consequently risky - not sure to what extent) special casing, trying to identify that precise error and ignoring the in-error status, relying on calling code to try the corresponding rollback and somehow ignoring the status for subsequent rollbacks to savepoints as well. 
- Rewriting these particular tests in such a way that they don't rely on savepoints to recover from database misuse

An alternate track (if it works) could maybe be to always call the save points rollbacks (not flag top level as broken but just low level as broken), but it looks like that would have a cost on production applications that may legitimately hit serialization failures. Unless these can be recovered by savepoints (can they, sometimes?) I would tend to not damage regular use in order to allow better workflow in testing database-misuse (your case IIUC).

**Comment by jtgeibel** [comment]
For additional context, our use case looks like this:

```rust
let tokens = api_tokens
    .filter(revoked.eq(false))
    .filter(token.eq(&token_));
conn.transaction(|conn| {
    update(tokens)
        .set(last_used_at.eq(now.nullable()))
        .get_result(conn)
})
.or_else(|_| tokens.first(conn))
```

First we try to obtain the token while updating the `last_used_at` time, but if that fails then we basically rerun the same query but as a `SELECT`. The connection could be read-only because the primary database is offline, or we could be doing maintenance. I did a bit of searching and didn't find another way to determine if the connection is read-only, and we do want test coverage to make sure this is working correctly.

Fortunately, I think we may be able to drop this fallback. We've reduced the endpoints that are allowed to authenticate with an API token and in the 1 remaining case where this fallback helps, it should be fine to just reject the requests when operating in a degraded mode. Though I don't think the general approach is that extreme and it would be nice if Diesel supported it.

> Diesel 1 had poor tracking of transactions failing, which sometimes led to Diesel thinking the depth was a certain level when in fact it was at a different level

I'm slowly working my way though some of the relevant logic, but I'm confused by this. It seems that is how [this logic](https://github.com/diesel-rs/diesel/blob/a02fcfbf57311a220e69dab03dd17e65a61313e9/diesel/src/connection/transaction_manager.rs#L342-L360) behaves, decreasing diesel's internal count without issuing the rollback. I don't know if it would work for the serialization use case, but perhaps we could attempt the rollback first and then test the connection status.

**Comment by Ten0** [comment]
> I'm confused by this

The idea there was that since "top level requires rollback" to fix the transaction, there's no point in actually casting all the intermediate rollback: these are only delaying the single rollback we need to do before trying the transaction again. By definition in that case the state after intermediate rollback is equivalent to the state before, so in that framework we can say it is correct to change transaction depth without actually interacting with the database. Of course as you pointed out in your case we consider that top level requires rollback when that is not the case.

**Comment by Ten0** [comment]
>  I did a bit of searching and didn't find another way to determine if the connection is read-only

Looks like this information could be added to:
https://github.com/rust-lang/crates.io/blob/1ba87040f9e5cbcce3ced57b43069a4788a65495/src/db.rs#L13
and
https://github.com/rust-lang/crates.io/blob/1ba87040f9e5cbcce3ced57b43069a4788a65495/src/db.rs#L139
when
https://github.com/rust-lang/crates.io/blob/08d6a5c74931954419bf82418920e41b2109dfdd/src/app.rs#L111-L167
is configured, no?

This would even prevent the failed attempt to write to the read-only database.

**Comment by weiznich** [comment]
@jtgeibel Are you interested in a `2.0.3` patch release containing this fix?

**Comment by jtgeibel** [comment]
@weiznich and @Ten0 thanks for investigating and landing the fix! I've tested against master and everything passes! A `2.0.3` release would be great.

**Comment by weiznich** [comment]
@jtgeibel I've just published 2.0.3 which contains a fix for this issue.
